### PR TITLE
Enable virtual time for generative agents

### LINKED ISSUE #4165: Enable virtual time in Generative Agents
### Feature request

Add extra input to the components of generative agents to enable virtual time instead of wall time

### Motivation

Because generative agents can "live in another world", it makes sense to enable virtual time

### Your contribution

I can submit a PR, in which I modified everything related to `datetime.now()`.

### PR DESCRIPTION
**Pull request by ifsheldon**
This should close https://github.com/hwchase17/langchain/issues/4165 and it should not cause any backward incompatibility issues.

### CODE REVIEW

File: langchain/experimental/generative_agents/generative_agent.py [hunk 1]
```diff
@@ -33,7 +33,7 @@ class GenerativeAgent(BaseModel):
     summary_refresh_seconds: int = 3600  #: :meta private:
     """How frequently to re-generate the summary."""
 
-    last_refreshed: datetime = Field(default_factory=datetime.now)  # : :meta private:
+    last_refreshed: datetime  # : :meta private:
```
Comment (vowelparrot) [inline langchain/experimental/generative_agents/generative_agent.py:36]: This could still have a default, right?

**Review by vowelparrot — COMMENTED** [review]
Overall looks good- could you run the notebook again and paste an output showing that it still works with these changes ?/ Thanks!

File: langchain/experimental/generative_agents/generative_agent.py (hunk 1 above)
Comment (ifsheldon) [reply langchain/experimental/generative_agents/generative_agent.py:36]: yeah. For backward compatibility, yes.

**Comment by ifsheldon** [comment]
@vowelparrot Here is the output exported as a PDF of the notebook. Nothing breaks.
[characters.pdf](https://github.com/hwchase17/langchain/files/11418990/characters.pdf)

**Comment by ifsheldon** [comment]
Any advice on fixing the lint failure? There is some type mismatch because I saved a datetime to the `outputs` dict of `GenerativeAgentMemory.save_context`, which has the type annotated as `Dict[str, str]`.

**Review by vowelparrot — APPROVED** [review]

**Comment by vowelparrot** [comment]
> Any advice on fixing the lint failure? There is some type mismatch because I saved a datetime to the `outputs` dict of `GenerativeAgentMemory.save_context`, which has the type annotated as `Dict[str, str]`.

You can update the dict typing to `Dict[str, Any]` in this case

**Review by hwchase17 — COMMENTED** [review]
lets do this without changing the get_relevant_documents signature

we can do this by mocking out datetime.now in a context manager

I can take over

**Pull request merged**